### PR TITLE
Support for MSC2285 (hidden read receipts)

### PR DESCRIPTION
**Pull request by SimonBrandner**
* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#changelog).
* [x] Pull request includes a [sign off](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#sign-off)
* [x] Code style is correct (run the [linters](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#code-style))

<hr>

Based on https://github.com/matrix-org/synapse/pull/5990
Implementation of https://github.com/matrix-org/matrix-doc/pull/2285

<hr>

To be honest, this PR doesn't fill me with feelings of confidence as I don't have much backend and Python development experience :sweat_smile: In case it's bad, sorry and feel free to reject it

### CODE REVIEW

File: synapse/handlers/receipts.py
Comment (SimonBrandner) [inline synapse/handlers/receipts.py:193]: This isn't exactly nice. Is preferable to use a prefixed key in the RR data or should we be doing this?

File: synapse/handlers/receipts.py
Comment (SimonBrandner) [inline synapse/handlers/receipts.py:193]: This isn't exactly nice. Is it preferable to use a prefixed key in the RR data or should we be doing this?

File: synapse/rest/client/v2_alpha/receipts.py
Comment (babolivier) [inline synapse/rest/client/v2_alpha/receipts.py:52]: Wrong parameter name.

File: synapse/rest/client/v2_alpha/receipts.py
Comment (babolivier) [inline synapse/rest/client/v2_alpha/receipts.py:51]: I'd just use the integer 400 here, for consistency with the rest of the code base.

File: synapse/rest/client/v2_alpha/read_marker.py
Comment (babolivier) [inline synapse/rest/client/v2_alpha/read_marker.py:47]: Same.

**Review by babolivier — CHANGES_REQUESTED** [review]
Thanks! Looks sane to me apart from a few things. Also I think we'll probably want to hide this feature behind a flag in the `experimental_features` config until the MSC merges.

**Comment by SimonBrandner** [comment]
Thanks for the review and sorry for the mistakes, hopefully, they're resolved now :sweat_smile:

File: tests/handlers/test_receipts.py
Comment (babolivier) [inline tests/handlers/test_receipts.py:289]: I think it'd be better to have it on the top of the file, and as a comment (`# foo`) block rather than a docstring.

File: synapse/handlers/receipts.py
Comment (babolivier) [inline synapse/handlers/receipts.py:160]: I think it would be better to use something like:

```suggestion
        if self.federation_sender and not (
            self.hs.config.experimental.msc2285_enabled and hidden
        ):
```

Otherwise it makes the condition a bit too difficult to read.

**Review by babolivier — CHANGES_REQUESTED** [review]
Looks really good now! Just a couple of bits and it should be good to go!

> sorry for the mistakes, hopefully, they're resolved now

Absolutely no worries at all, thanks a lot for working on this contribution! :slightly_smiling_face:

File: tests/handlers/test_receipts.py
Comment (babolivier) [inline tests/handlers/test_receipts.py:26]: I'm going to be annoying again (sorry :sweat_smile:) but can you just put it between `prepare` and `test_filters_out_hidden_receipt`? The intention is that it's as fresh as possible in the mind of people when they start reading the actual tests. Sorry I wasn't clear enough!

**Review by babolivier — CHANGES_REQUESTED** [review]
(this is the last one I promise)

File: tests/handlers/test_receipts.py
Comment (SimonBrandner) [reply tests/handlers/test_receipts.py:26]: Absolutely no worries! :smile:

**Review by babolivier — APPROVED** [review]
Looks good to me! I'll just wait for the CI to complete and merge this. Thanks again for your contribution! :smile:

File: synapse/handlers/receipts.py
Comment (babolivier) [reply synapse/handlers/receipts.py]: Just to be sure; I assume my comment about using a constant answered your question here?

Comment (SimonBrandner) [reply synapse/handlers/receipts.py]: Yep

File: changelog.d/10413.feature
Comment (babolivier) [inline changelog.d/10413.feature:1]: (don't mind this, I'm just applying this myself so you're mentioned in the changelog for this; feel free to add it yourself next time :slightly_smiling_face:)

```suggestion
Support for [MSC2285 (hidden read receipts)](https://github.com/matrix-org/matrix-doc/pull/2285). Contributed by @SimonBrandner.
```

**Comment by Bugsbane** [comment]
So, does this merge support:
1) hiding read receipts for anyone in a specific room
2) showing read receipts from other users in the room, or 
3) hiding your read receipts to others in the same room?

I'm running 1.59.7 (obviously after this merge was accepted) and I still don't see any way of avoiding having read receipts hidden in clients using rooms on my Synapse server.

**Comment by tio-trom** [comment]
> this feature is hidden behind a labs option as it is still considered unstable (the MSC isn't merged yet)

I cannot find this in the labs section how can I enable it? I'd like to test it. Thanks! :)

**Comment by SimonBrandner** [comment]
> > this feature is hidden behind a labs option as it is still considered unstable (the MSC isn't merged yet)
> 
> I cannot find this in the labs section how can I enable it? I'd like to test it. Thanks! :)

I would recommend visiting https://matrix.to/#/#synapse:matrix.org :)

**Pull request merged**
